src/document.h

```diff
@@ -1117,12 +1111,31 @@ struct Document {
                 if (sys->frame->borddd) sys->frame->borddd->ShowPopup();
                 break;
 
+            case A_REPLACEONCE:
+            case A_REPLACEONCEJ:
             case A_REPLACEALL: {
                 if (!sys->searchstring.Len()) return _(L"No search.");
-                rootgrid->AddUndo(this);  // expensive?
-                rootgrid->FindReplaceAll(sys->frame->replaces->GetValue());
-                rootgrid->ResetChildren();
-                Refresh();
+                if (k == A_REPLACEALL) {
+                    wxString filter = sys->frame->filter->GetValue();
+                    wxString replaces = sys->frame->replaces->GetValue();
+                    rootgrid->AddUndo(this);  // expensive?

```

Comment:
Yes that is going to copy the whole document. Since there is a cap on the amount of memory undoes can hold, the bad effect of this is that it may clear the rest of the undo history.

The alternative is a single undo per replaced cell. That is likely ok if its just a few replacements but annoying for a large amount..

Better yet would be complicated: first find where all the replacements will take place, then create an undo only for that area, then do the replacement